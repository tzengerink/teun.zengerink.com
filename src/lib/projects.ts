import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import yaml from 'js-yaml'

export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}
export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug, `${Width.Mobile}w`)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      exports: MAX_WIDTHS.map((width) => ({ width, url: `/photos/${slug}/${width}w/${filename}` })),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      isArchived: project.isArchived ?? false,
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  let projects: Project[] = []

  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    projects = await generateProjects(config)
  } catch (e) {
    console.error(e)
  }

  return projects
}
