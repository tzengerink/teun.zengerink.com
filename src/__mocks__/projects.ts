import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The *statement* for the _project_',
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/01.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/02.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/03.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/03.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        caption: 'This is _image 1_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/01.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        caption: 'This is _image 2_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/02.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        caption: 'This is _image 3_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/03.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/03.jpg` },
        ],
      },
      {
        key: '04',
        caption: 'This is _image 4_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/04.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/04.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Archived project',
    slug: 'archived-project',
    isArchived: true,
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/01.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/02.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/02.jpg` },
        ],
      },
    ],
  },
]

export default projects
