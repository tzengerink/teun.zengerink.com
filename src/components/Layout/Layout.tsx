import classNames from 'classnames'
import Head from 'next/head'
import React from 'react'
import { Project } from '../../lib/projects'
import Navigation from '../Navigation/Navigation'

const TITLE_ROOT = 'Teun Zengerink'
const TITLE_DIV = '-'

interface LayoutProps {
  title?: string
  projects: Project[]
  children?: React.ReactNode
}

const Layout: React.FC<LayoutProps> = ({ title, projects, children }) => {
  const pageTitle = React.useMemo(() => (title ? [TITLE_ROOT, title].join(` ${TITLE_DIV} `) : TITLE_ROOT), [title])

  return (
    <div>
      <Head>
        <meta charSet="utf-8" />
        <meta name="description" content={pageTitle}></meta>
        <title>{pageTitle}</title>
        <link rel="icon" href="/favicon.ico" />
        <link
          href="https://fonts.googleapis.com/css2?family=Roboto+Mono:ital,wght@0,300;0,500;1,300&display=swap"
          rel="stylesheet"
        ></link>
      </Head>
      <Navigation pageTitle={TITLE_ROOT} projects={projects} />
      <main className={classNames('relative', 'mt-xl', 'mx-md', 'mb-md', 'md:absolute', 'md:m-0', 'md:inset-0')}>
        {projects?.length ? children : null}
      </main>
    </div>
  )
}

export default Layout
