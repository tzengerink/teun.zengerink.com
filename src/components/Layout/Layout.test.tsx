import { render } from '@testing-library/react'
import React from 'react'
import projects from '../../__mocks__/projects'
import Layout from './Layout'

jest.mock('../Navigation/Navigation', () => ({
  __esModule: true,
  default: () => <div>[NAVIGATION]</div>,
}))

describe('Layout', () => {
  it('renders nothing when initializing', () => {
    const { container } = render(<Layout projects={null} />)
    expect(container).toMatchSnapshot()
  })

  it('renders the children when initialized', () => {
    const child = <div>Hello World!</div>
    const { container } = render(<Layout projects={projects}>{child}</Layout>)
    expect(container).toMatchSnapshot()
  })
})
