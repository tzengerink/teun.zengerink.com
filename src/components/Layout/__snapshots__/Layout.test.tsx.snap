// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders nothing when initializing 1`] = `
<div>
  <div>
    <div>
      [NAVIGATION]
    </div>
    <main
      class="relative mt-xl mx-md mb-md md:absolute md:m-0 md:inset-0"
    />
  </div>
</div>
`;

exports[`Layout renders the children when initialized 1`] = `
<div>
  <div>
    <div>
      [NAVIGATION]
    </div>
    <main
      class="relative mt-xl mx-md mb-md md:absolute md:m-0 md:inset-0"
    >
      <div>
        Hello World!
      </div>
    </main>
  </div>
</div>
`;
